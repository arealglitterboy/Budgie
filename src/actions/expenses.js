import {v4 as uuid} from 'uuid';

export const addExpense = (
    {
        description = '',
        note = '',
        date = 0,
        currency = 'â‚¬',
        amount = 0
    } = {}
    ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        date,
        currency,
        amount
    }
});

export const removeExpense = ({ id }) => ({ type: 'REMOVE_EXPENSE', id });

export const editExpense = (id, updates) => ({ type: 'EDIT_EXPENSE', id, updates });