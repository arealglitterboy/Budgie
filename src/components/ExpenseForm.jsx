import React, { useState } from 'react';
import { connect } from 'react-redux';

import ReactDatePicker from 'react-datepicker';
import TextareaAutosize from 'react-textarea-autosize';

import {Input} from './Input';
import InputCreateSelect from './InputCreateSelect';
import InputMultiSelect from './InputMultiSelect';
import {InputArea} from './InputArea';

import { categoryOptions } from './filter/options';
import { addNewContact } from '../actions/contacts.action';

const isValidAmount = (amount) => (!amount || amount.match(/^(0|[1-9]\d*)(\.\d{0,2})?$/gm));

const findErrors = (title='', contact, date, amount) => ({
    title: !title,
    contact: !(contact && contact.name),
    date: !date,
    amount: !Number.isFinite(amount)
});

// TODO: Finish this
export const ExpenseForm = ({expense={}, ...props}) => {
    const [title, setTitle] = useState(expense.title || '');
    const [contact, setContact] = useState(expense.contact || { id: '', name: '' });
    const [categories, setCategories] = useState(expense.categories || []);
    const [note, setNote] = useState(expense.note || '');
    const [date, setDate] = useState(expense.date ? new Date(expense.date) : new Date());
    const [amount, setAmount] = useState(expense.amount ? (expense.amount/100).toFixed(2) : 0);
    const [error, setError] = useState({});

    const onSubmit = (e) => {
        e.preventDefault();

        setError(findErrors(title, contact, date, amount));

        if (Object.keys(error).length === 0) {
            if (contact.name === contact.id) {
                setContact(props.addContact({ name: contact.name, date: new Date() }).contact);
            }
            props.onSubmit({ contact: contact.id, title, note, categories, date, amount: Math.floor(amount.replace(',', '.') * 100) });
        }
    };

    const DateInput = React.forwardRef(({ value, onClick }, ref) => <Input type="text" label="Date" id="set-date" ref={ref} value={value} onChange={setDate} onClick={onClick} />);

    return (
        <>
            <form action="" onSubmit={onSubmit} className="expense-form">
                <TextareaAutosize className="expense-form__title" id='title' placeholder='Title' onChange={({ target }) => setTitle(target.value)} value={title} />

                <fieldset className="expense-form__date">
                    {error.date && <p className="expense-form__error">Invalid date</p>}
                    <ReactDatePicker
                        selectsStart
                        id="date-picker"
                        className="date-picker expense-form__date__input"
                        dateFormat="dd/MM/yyyy"

                        selected={date}
                        onChange={setDate}
                        customInput={<DateInput />}
                    />
                </fieldset>

                <fieldset className="expense-form__amount">
                    {error.amount && <p className="expense-form__error">Invalid amount</p>}
                    <Input
                        className="expense-form__amount"
                        type="text"
                        id='amount'
                        label="Amount"
                        value={amount}
                        onChange={setAmount}
                        validator={isValidAmount}
                    />
                </fieldset>

                <fieldset className="expense-form__contact">
                    {error.contact && <p className="expense-form__error">Invalid contact</p>}
                    <InputCreateSelect
                        value={{label: contact.name, value: contact.id}}
                        onChange={({value, label}) => setContact({id: value, name: label})}
                        id='contact'
                        label='Contact'
                        options={props.contacts.map(({id, name}) => ({ label: name, value: id }))}
                    />
                </fieldset>

                <fieldset className="expense-form__categories">
                    {error.categories && <p className="expense-form__error">Invalid categories</p>}
                    <InputMultiSelect
                        value={categories.map(category => categoryOptions.find(({value}) => value === category))}
                        className="expense-form__categories"
                        type="text"
                        label="Categories"
                        id='categories'
                        onChange={setCategories}
                        options={categoryOptions}
                    />
                </fieldset>

                <fieldset className="expense-form__note">
                    <InputArea
                        id="note"
                        label="Note"
                        onChange={setNote}
                        placeholder="What were the details"
                        value={note}
                    />
                </fieldset>

                <button className="expense-form__submit" type="submit">Confirm</button>
            </form>
        </>
    );
};

// export class ExpenseForm extends React.Component {
//     state = {
//         title: this.props.expense ? this.props.expense.title : '',
//         contact: this.props.expense ? { id, name } = this.props.expense.contact : { id: undefined, name: undefined },
//         categories: this.props.expense ? this.props.expense.categories : [],
//         note: this.props.expense ? this.props.expense.note : '',
//         date: this.props.expense ? new Date(this.props.expense.date) : new Date(),
//         amount: this.props.expense ? (this.props.expense.amount/100).toFixed(2).toString() : '',
//         error: ''
//     };
    
//     onSubmit = (e) => {
//         e.preventDefault();

//         let error = '';
//         const { contact, title, note, amount, categories, date } = this.state;

//         if (!(title.trim() && amount && (contact && contact.name) && date)) {
//             error= `ERROR: Nice try, you need to fill out all non-optional fields`;
//         } else {
//             if (contact.name === contact.id) {
//                 const newContact = this.props.addContact({ name: contact.name, date: new Date() }).contact;

//                 contact.id = newContact.id;
//                 contact.date = newContact.date;
//             }
//             this.props.onSubmit({ contact: contact.id, title, note, categories, date, amount: Math.floor(amount.replace(',', '.') * 100) });
//         }

//         this.setState(() => ({ error }));
//     }

//     isValidAmount = (amount => !amount || amount.match(/^(0|[1-9]\d*)(\.\d{0,2})?$/gm));

//     onContactChange = ({ value:id='', label:name='' } = {}) => {
//         this.setState(() => ({ contact: { id, name } }));
//     };

//     onTitleChange = (title) => {this.setState(() => ({ title }))};

//     onNoteChange = (note) => {this.setState(() => ({ note }))};

//     onCategoriesChange = (e) => {
//         const categories = e.target.value;
//         this.setState({ categories });
//     };

//     onAmountChange = (amount) => {
//         if (this.isValidAmount(amount)) {
//             this.setState(() => ({ amount }));
//         }
//     };

//     onClickCloseDialog = (e) => {
//         e.preventDefault();
//         this.setState(() => ({ error: '' }));
//     };

//     setDate = (date) => {
//         this.setState(() => ({ date }));
//     }

//     render() {
//         const style='secondary';
//         const DateInput = React.forwardRef(({ value, onClick }, ref) => <Input data-style={style} type="text" label="Date" id="set-date" ref={ref} value={value} onChange={this.setDate} onClick={onClick} />);

//         return (
//             <>
//                 <form action="" onSubmit={this.onSubmit} className="expense-form">

//                     <TextareaAutosize className="expense-form__title" id='title' placeholder='Title' onChange={(e) => this.onTitleChange(e.target.value)} value={this.state.title} />

//                     <fieldset className="expense-form__date">
//                         <ReactDatePicker
//                             selectsStart
//                             id="date-picker"
//                             className="date-picker expense-form__date__input"
//                             dateFormat="dd/MM/yyyy"

//                             selected={this.state.date}
//                             onChange={this.setDate}
//                             customInput={<DateInput />}
//                         />
//                     </fieldset>

//                     <fieldset className="expense-form__amount">
//                         <Input
//                             data-style={style}
//                             className="expense-form__amount"
//                             type="text"
//                             id='amount'
//                             label="Amount"
//                             value={this.state.amount}
//                             onChange={this.onAmountChange}
//                             validator={this.isValidAmount}
//                         />
//                     </fieldset>

//                     <fieldset className="expense-form__contacts">
//                         <InputCreateSelect
//                             value={{label: this.state.contact.name, value: this.state.contact.id}}
//                             onChange={this.onContactChange}
//                             id='contact'
//                             label='Contact'
//                             options={this.props.contacts.map(({id: value, name: label}) => ({ label, value }))}
//                         />
//                     </fieldset>

//                     <fieldset className="expense-form__categories">
//                         <Input
//                             data-style={style}
//                             className="expense-form__categories"
//                             type="text"
//                             label="Categories"
//                             id='categories'
//                             value={this.state.categories.toString()}
//                             onChange={this.onCategoriesChange}
//                         />
//                     </fieldset>

//                     <fieldset className="expense-form__note">
//                         <InputArea data-style={style} id="note" label="Note" onChange={this.onNoteChange} placeholder="What were the details" value={this.state.note} />
//                     </fieldset>

//                     <button className="expense-form__submit" type="submit">Confirm</button>
//                 </form>
//                 {
//                     (this.state.error) &&
//                     <footer className="expense-form__error-box" open>
//                         <h3>Warning!</h3>
//                         <p>{this.state.error}</p>
//                         <button value="cancel" className="expense-form__error-box__close" onClick={this.onClickCloseDialog}>Close</button>
//                     </footer>
//                 }
//             </>
//         );
//     }
// }

const mapDispatchToProps = (dispatch) => ({
    addContact: (contact) => dispatch(addNewContact(contact))
});

const mapStateToProps = connect(({ contacts }) => ({ contacts }), mapDispatchToProps);

export default mapStateToProps(ExpenseForm);