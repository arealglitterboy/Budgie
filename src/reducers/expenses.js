const defaultState = [];

export default (state = defaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((e) => (e.id !== action.id));
        case 'EDIT_EXPENSE': // ? Map each value in the array into a new array, if the id matches the selected id map it with its new values
            return state.map((e) => ((e.id === action.id) ? {...e, ...action.updates} : {...e}));
        default:
            return state;
    }
};