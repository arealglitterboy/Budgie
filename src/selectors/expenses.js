import moment from "moment";

const includesIgnoreCase = (key, searching) => (!key || (searching && searching.toLowerCase().includes(key.toLowerCase())));

/**
 * 
 * @param {string} search 
 * @returns {function}
 */
const createIncludes = (search) => {
    const regExp = new RegExp(search, 'gi')
    return (str) => str.search(regExp) >= 0;
};

function filterExpense({ description, note, date }, term, startDate, endDate) {
    const includes = createIncludes(term);
    const dateMoment = moment(date);

    const containsTerm = includes(description) || includes(note); // ? If either the description or the note contains the search term.
    // const containsTerm = includesIgnoreCase(term, description) || includesIgnoreCase(term, note); // ? If either the description or the note contains the search term.
    const inDate = (!startDate || startDate.isSameOrBefore(dateMoment, 'day')) && (!endDate || endDate.isSameOrAfter(dateMoment, 'day')); // ? If the start/end date is defined, and the given date is greater/less than it.
    return containsTerm && inDate;
}

export default (expenses, { term, startDate, endDate, sortFunction }) => {
    return expenses.filter((e) => (filterExpense(e, term, startDate, endDate))).sort(sortFunction);
};