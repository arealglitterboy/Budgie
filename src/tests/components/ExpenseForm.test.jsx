import React from "react";
import { shallow } from "enzyme";

import ExpenseForm from "../../components/ExpenseForm";
import { expenses } from "../fixtures/expenses.fixture";

const getWrapper = (props = {}) => (shallow(<ExpenseForm {...props} />));

const simulateChange = (id, state, value, wrapper = getWrapper()) => {
    wrapper.find(`#${id}`).simulate('change', { target: { value } });
    expect(wrapper.state(state)).toBe(value);
    return wrapper;
};

test('should render expense form correctly', () => {
    // const wrapper = shallow(<ExpenseForm />);
    const wrapper = getWrapper({ today: new Date("2021-09-20T16:00:41.960Z") });
    expect(wrapper).toMatchSnapshot();
});

test('should render expense form with expense data', () => {
    const wrapper = getWrapper({ expense: expenses[0] });
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
    const wrapper = getWrapper({ today: new Date("2021-09-20T16:00:41.990Z") });
    expect(wrapper).toMatchSnapshot(); // Check before the error

    wrapper.find('form').simulate('submit', { preventDefault: () => {} }); // Find the form, then simulate a submit event with an event object
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot(); // Check after the error
});

test('should set description on input change', () => {
    simulateChange('description', 'description', "description change");
});

test('should set note on text area change', () => {
    simulateChange('note', 'note', "note change");
});

test('should set amount if valid input', () => {
    simulateChange('amount', 'amount', '23.50');
})

test('should not set amount if not valid input', () => {
    const value = '23.50';
    const wrapper = simulateChange('amount', 'amount', value);
    wrapper.find('#amount').simulate('change', { target: { value: '23.50' } }); // Test/set amount with a valid amount
    expect(wrapper.state('amount')).toBe(value);
    
    wrapper.find('#amount').simulate('change', { target: { value: '-14.45' } }); // Test an invalid amount
    expect(wrapper.state('amount')).toBe(value);
    
    wrapper.find('#amount').simulate('change', { target: { value: "Hello World" } }); // Test an invalid value
    expect(wrapper.state('amount')).toBe(value);
});

test('should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = getWrapper({ expense: expenses[0], onSubmit: onSubmitSpy });
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });
    expect(wrapper.state('error')).toBe('');
    const { description, amount, note, date, currency } = expenses[0];
    expect(onSubmitSpy).toHaveBeenLastCalledWith({ description, amount, note, date, currency });
});

test('should set new date on date changed', () => {
    const wrapper = getWrapper();
    const now = new Date();
    wrapper.find('#date-picker').prop('onChange')(now);
    expect(wrapper.state('date')).toEqual(now);
});