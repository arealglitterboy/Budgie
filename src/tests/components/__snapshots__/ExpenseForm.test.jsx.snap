// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render error for invalid form submission 1`] = `
<section>
  <form
    action=""
    className="expense-form"
    onSubmit={[Function]}
  >
    <ForwardRef(TextareaAutosize)
      className="expense-form__title"
      id="title"
      onChange={[Function]}
      placeholder="Title"
      value=""
    />
    <fieldset
      className="expense-form__date"
    >
      <r
        allowSameDay={false}
        className="date-picker expense-form__date__input"
        customInput={<ForwardRef />}
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        id="date-picker"
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2021-12-21T11:15:56.996Z}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </fieldset>
    <fieldset
      className="expense-form__amount"
    >
      <Input
        className="expense-form__amount"
        data-style="secondary"
        id="amount"
        label="Amount"
        onChange={[Function]}
        type="text"
        validator={[Function]}
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__contacts"
    >
      <_default
        id="contact"
        label="Contact"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Mark Matthews",
              "value": "124",
            },
            Object {
              "label": "Alex Vincent",
              "value": "1642",
            },
            Object {
              "label": "Bank Of Ireland",
              "value": "764",
            },
          ]
        }
        value={
          Object {
            "id": undefined,
            "name": undefined,
          }
        }
      />
    </fieldset>
    <fieldset
      className="expense-form__categories"
    >
      <Input
        className="expense-form__categories"
        data-style="secondary"
        id="categories"
        label="Categories"
        onChange={[Function]}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__note"
    >
      <InputArea
        data-style="secondary"
        id="note"
        label="Note"
        onChange={[Function]}
        placeholder="What were the details"
        value=""
      />
    </fieldset>
    <button
      className="expense-form__submit"
      type="submit"
    >
      Confirm
    </button>
  </form>
</section>
`;

exports[`should render error for invalid form submission 2`] = `
<section>
  <form
    action=""
    className="expense-form"
    onSubmit={[Function]}
  >
    <ForwardRef(TextareaAutosize)
      className="expense-form__title"
      id="title"
      onChange={[Function]}
      placeholder="Title"
      value=""
    />
    <fieldset
      className="expense-form__date"
    >
      <r
        allowSameDay={false}
        className="date-picker expense-form__date__input"
        customInput={<ForwardRef />}
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        id="date-picker"
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2021-12-21T11:15:56.996Z}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </fieldset>
    <fieldset
      className="expense-form__amount"
    >
      <Input
        className="expense-form__amount"
        data-style="secondary"
        id="amount"
        label="Amount"
        onChange={[Function]}
        type="text"
        validator={[Function]}
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__contacts"
    >
      <_default
        id="contact"
        label="Contact"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Mark Matthews",
              "value": "124",
            },
            Object {
              "label": "Alex Vincent",
              "value": "1642",
            },
            Object {
              "label": "Bank Of Ireland",
              "value": "764",
            },
          ]
        }
        value={
          Object {
            "id": undefined,
            "name": undefined,
          }
        }
      />
    </fieldset>
    <fieldset
      className="expense-form__categories"
    >
      <Input
        className="expense-form__categories"
        data-style="secondary"
        id="categories"
        label="Categories"
        onChange={[Function]}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__note"
    >
      <InputArea
        data-style="secondary"
        id="note"
        label="Note"
        onChange={[Function]}
        placeholder="What were the details"
        value=""
      />
    </fieldset>
    <button
      className="expense-form__submit"
      type="submit"
    >
      Confirm
    </button>
  </form>
  <footer
    className="expense-form__error-box"
    open={true}
  >
    <h3>
      Warning!
    </h3>
    <p>
      ERROR: Nice try, you need to fill out all non-optional fields
    </p>
    <button
      className="expense-form__error-box__close"
      onClick={[Function]}
      value="cancel"
    >
      Close
    </button>
  </footer>
</section>
`;

exports[`should render expense form correctly 1`] = `
<section>
  <form
    action=""
    className="expense-form"
    onSubmit={[Function]}
  >
    <ForwardRef(TextareaAutosize)
      className="expense-form__title"
      id="title"
      onChange={[Function]}
      placeholder="Title"
      value=""
    />
    <fieldset
      className="expense-form__date"
    >
      <r
        allowSameDay={false}
        className="date-picker expense-form__date__input"
        customInput={<ForwardRef />}
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        id="date-picker"
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2021-12-21T11:15:56.557Z}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </fieldset>
    <fieldset
      className="expense-form__amount"
    >
      <Input
        className="expense-form__amount"
        data-style="secondary"
        id="amount"
        label="Amount"
        onChange={[Function]}
        type="text"
        validator={[Function]}
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__contacts"
    >
      <_default
        id="contact"
        label="Contact"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Mark Matthews",
              "value": "124",
            },
            Object {
              "label": "Alex Vincent",
              "value": "1642",
            },
            Object {
              "label": "Bank Of Ireland",
              "value": "764",
            },
          ]
        }
        value={
          Object {
            "id": undefined,
            "name": undefined,
          }
        }
      />
    </fieldset>
    <fieldset
      className="expense-form__categories"
    >
      <Input
        className="expense-form__categories"
        data-style="secondary"
        id="categories"
        label="Categories"
        onChange={[Function]}
        type="text"
        value=""
      />
    </fieldset>
    <fieldset
      className="expense-form__note"
    >
      <InputArea
        data-style="secondary"
        id="note"
        label="Note"
        onChange={[Function]}
        placeholder="What were the details"
        value=""
      />
    </fieldset>
    <button
      className="expense-form__submit"
      type="submit"
    >
      Confirm
    </button>
  </form>
</section>
`;

exports[`should render expense form with expense data 1`] = `
<section>
  <form
    action=""
    className="expense-form"
    onSubmit={[Function]}
  >
    <ForwardRef(TextareaAutosize)
      className="expense-form__title"
      onChange={[Function]}
      placeholder="Title"
      value="Food"
    />
    <fieldset
      className="expense-form__date"
    >
      <r
        allowSameDay={false}
        className="date-picker expense-form__date__input"
        customInput={<ForwardRef />}
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        id="date-picker"
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selected={2021-04-09T00:00:00.000Z}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </fieldset>
    <fieldset
      className="expense-form__amount"
    >
      <Input
        className="expense-form__amount"
        data-style="secondary"
        label="Amount"
        onChange={[Function]}
        type="text"
        validator={[Function]}
        value="10.50"
      />
    </fieldset>
    <fieldset
      className="expense-form__contacts"
    >
      <Input
        data-style="secondary"
        label="Contact"
        onChange={[Function]}
        type="text"
        value={124}
      />
    </fieldset>
    <fieldset
      className="expense-form__categories"
    >
      <Input
        className="expense-form__categories"
        data-style="secondary"
        label="Categories"
        onChange={[Function]}
        type="text"
        value="Food"
      />
    </fieldset>
    <fieldset
      className="expense-form__note"
    >
      <InputArea
        data-style="secondary"
        id="note"
        label="Note"
        onChange={[Function]}
        placeholder="What were the details"
        value="Bought food for mastication"
      />
    </fieldset>
    <button
      className="expense-form__submit"
      type="submit"
    >
      Confirm
    </button>
  </form>
</section>
`;
